### ISSUE TITLE
Accessibility, macOS: Pressing X does not automatically speak

### ISSUE THREAD

**Issue by TheQuinbox** [nolanlawson/pinafore#2180]
I'm not sure if this is intentional, but it works in every other setup I've tried on Windows, and i actually really like it.

If you're using a screen reader, and press X to show/hide a content warning, the toot gets reread, with that applied. On the Mac, however, this does not happen, making me have to scroll over it again to read the toot.

macOS Ventura, latest Safari, and VoiceOver.

**Comment by MarcoZehe** [comment]
This is because the aria-label gets updated, but neither Chromium browsers nor Safari fire an appropriate Mac platform event for this. On Windows, a name change event gets fired on the focused item which NVDA picks up on, and JAWS sometimes does too if it feels like it. I [had this working](https://bugzilla.mozilla.org/show_bug.cgi?id=1682247) in Firefox for Mac nicely, but it caused a [different problem](https://bugzilla.mozilla.org/show_bug.cgi?id=1749063) and got reverted later, so Firefox also doesn't notify Voiceover of the name change event any more. The funny thing is that VoiceOver reacted positively to the AXTitleChangedEvent on Pinafore, even though Safari and Chromium don't even expose AXTitle on Article elements. Firefox did, until it was removed ten months ago to be on par with Safari. Which means that it now has the same bug in Pinafore as Safari and Chrome/Edge do. AFAIK, there is nothing we can do from the app side of things to get this working. Browsers and/or VoiceOver need to change. Unfortunately, Apple are very inconsistent with what they expose when to VoiceOver. AKA there is no definitive documentation on when to use AXTitle or AXDescription, or both.

**Comment by nolanlawson** [comment]
Hm, there are sometimes hacks that the app author can do to get this to work correctly. E.g. remove the DOM node from the DOM and re-insert it. I don't know if it's worth it for Pinafore to try to work around browser/screenreader bugs though.

Thanks for the context @MarcoZehe !

**Comment by MarcoZehe** [comment]
Yeah, this might work, @nolanlawson, but it gets complicated by the fact that you'd be removing and reinserting the focused node for this. Or do you mean only the part that contains the text, and not the other information?

**Comment by nolanlawson** [comment]
@MarcoZehe Ah you're right, this would get very tricky very fast. Also I have no evidence this would actually work; just a hunch based on behavior I've seen from screen readers in the past.

It's a shame that this doesn't work out-of-the-box. :confused:

**Comment by joel350** [comment]
Thanks

**Comment by pitermach** [comment]
How about using the live region that was recently added for the boost/favorite notifications to speak the post label when the CW is toggled? THis would probably have to be either tied to the user agent - do this only if someone is using a Mac, or a setting. I know of at least one Windows Screen Reader - ZDSR, which also doesn't react to the title change so it might be useful there as well.

Feel free to tell me how horrible of an idea this is

**Comment by jcsteh** [comment]
That would result in double speaking on browser/screen reader combinations which behave correctly, though. And using a live region alone would be bad for everyone, especially braille users.

**Comment by nolanlawson** [comment]
This sounds a bit brittle to me - I try to avoid UA sniffing as much as possible... Is there a bug filed on WebKit/VoiceOver somewhere?
